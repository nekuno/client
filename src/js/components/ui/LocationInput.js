import React, { PropTypes, Component } from 'react';
import Geosuggest from './Geosuggest';

export default class LocationInput extends Component {

    static propTypes = {
        placeholder    : PropTypes.string.isRequired,
        onSuggestSelect: PropTypes.func.isRequired,
        autoFocus      : PropTypes.bool
    };

    static defaultProps = {
        autoFocus: true
    };

    constructor() {
        super();

        this.onFocusHandler = this.onFocusHandler.bind(this);
        this.onSuggestSelect = this.onSuggestSelect.bind(this);
    }

    componentDidMount() {
        if (this.props.autoFocus) {
            this.refs.geosuggest.focus();
        }
    }

    onSuggestSelect(suggest) {
        let locality = '', country = '';
        suggest.gmaps.address_components.forEach(function(component) {
            component.types.forEach(function(type) {
                if (!locality && (type === 'locality' || type === 'administrative_area_level_4')) {
                    locality = component.long_name;
                }
                if (!country && type === 'country') {
                    country = component.long_name;
                }
            });
        });

        let location = {
            latitude : parseFloat(suggest.location.lat),
            longitude: parseFloat(suggest.location.lng),
            address  : suggest.label,
            locality : locality,
            country  : country
        };

        return this.props.onSuggestSelect(location);
    }

    render() {
        return (
            <div className="item-content">
                <div className="item-inner location-inner">
                    <div className="item-input" ref="geosuggestWrapper">
                        <Geosuggest {...this.props}
                            autoActivateFirstSuggest={true}
                            initialValue={this.props.defaultValue}
                            placeholder={this.props.placeholder}
                            ref="geosuggest"
                            onSuggestSelect={this.onSuggestSelect}
                            getSuggestLabel={function(suggest) { return suggest.description.length > 35 ? suggest.description.slice(0, 35) + '...' : suggest.description }}
                            onFocus={this.onFocusHandler}
                            skipSuggest={function(suggest) { return suggest.terms.length < 2 }}/>
                    </div>
                </div>
            </div>
        );
    }

    onFocusHandler() {
        /*let geosuggestWrapper = this.refs.geosuggestWrapper;
        window.setTimeout(function() {
            geosuggestWrapper.scrollIntoView();
            document.getElementsByClassName('view')[0].scrollTop -= 100;
        }, 500)*/
    }
}