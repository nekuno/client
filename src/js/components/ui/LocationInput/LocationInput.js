import PropTypes from 'prop-types';
import React, { Component } from 'react';
import Geosuggest from 'react-geosuggest';
import styles from './LocationInput.scss';
import Input from "../Input/Input";
import {action} from "@storybook/addon-actions";

export default class LocationInput extends Component {

    static propTypes = {
        title          : PropTypes.string,
        defaultValue   : PropTypes.string,
        placeholder    : PropTypes.string.isRequired,
        onSuggestSelect: PropTypes.func.isRequired,
        autoFocus      : PropTypes.bool,
        clearOnSelect  : PropTypes.bool
    };

    constructor() {
        super();

        this.onFocusHandler = this.onFocusHandler.bind(this);
        this.onSuggestSelect = this.onSuggestSelect.bind(this);
    }

    componentDidMount() {
        if (this.props.autoFocus) {
            this.refs.geosuggest.focus();
        }
    }

    onSuggestSelect(suggest) {
        let locality = '', country = '';
        if (suggest) {
            suggest.gmaps.address_components.forEach(function(component) {
                component.types.forEach(function(type) {
                    if (!locality && (type === 'locality' || type === 'administrative_area_level_4')) {
                        locality = component.long_name;
                    }
                    if (!country && type === 'country') {
                        country = component.long_name;
                    }
                });
            });

            let location = {
                latitude : parseFloat(suggest.location.lat),
                longitude: parseFloat(suggest.location.lng),
                address  : suggest.label,
                locality : locality,
                country  : country
            };

            if (this.props.clearOnSelect) {
                this.refs['geosuggest'].clear();
            }

            return this.props.onSuggestSelect(location);
        }
    }

    onFocusHandler() {
        let geosuggestWrapper = this.refs.geosuggestWrapper;
        if (geosuggestWrapper) {
            window.setTimeout(function () {
                geosuggestWrapper.scrollIntoView();
                document.getElementsByClassName('view')[0].scrollTop -= 100;
            }, 500);
        }
    }

    render() {
        const {title, defaultValue, placeholder} = this.props;

        const inputStyles = { 'input': {'test': 'test', 'anothertest': 'anothertest'}, 'suggests': {}, 'suggestItem': {} };

        return (
            <div className={styles.locationInputWrapper}>
                <div className={styles.title}>{title}</div>
                <div className={styles.locationInput} ref="geosuggestWrapper">
                    <Geosuggest {...this.props}
                                autoActivateFirstSuggest={true}
                                initialValue={defaultValue}
                                placeholder={placeholder}
                                inputClassName={'additional-class'}
                                ref="geosuggest"
                                onSuggestSelect={this.onSuggestSelect}
                                getSuggestLabel={function(suggest) { return suggest.description.length > 35 ? suggest.description.slice(0, 35) + '...' : suggest.description }}
                                onFocus={this.onFocusHandler}
                                skipSuggest={function(suggest) { return suggest.terms.length < 2 || !suggest.types.some(type => type === 'locality' || type === 'administrative_area_level_4') }}
                                required
                    />
                </div>
            </div>
        );
    }
}