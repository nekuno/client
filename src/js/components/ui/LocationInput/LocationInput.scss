@import '../../../../scss/_partials/colors';

.location-input-wrapper {

  .title {
    font-weight: 600;
  }
  .location-input {


  }
}