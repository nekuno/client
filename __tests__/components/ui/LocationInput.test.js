import React from 'react';
import LocationInput from '../../../src/js/components/ui/LocationInput/LocationInput.js';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('Test LocationInput component', () => {
    const mockFn = jest.fn();
    const locationInput = shallow(<LocationInput title={'Lorem ipsum'} placeholder={'Write a location'} onSuggestSelect={mockFn}/>);

    it('should be defined', () => {
        expect(locationInput).toBeDefined();
    });
    it('should render correctly', () => {
        expect(locationInput).toMatchSnapshot();
    });
});
