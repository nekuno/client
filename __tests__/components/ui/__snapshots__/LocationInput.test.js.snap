// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test LocationInput component should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LocationInput
    onSuggestSelect={[MockFunction]}
    placeholder="Write a location"
    title="Lorem ipsum"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="title"
        >
          Lorem ipsum
        </div>,
        <div
          className="locationInput"
        >
          <Geosuggest
            autoActivateFirstSuggest={true}
            bounds={null}
            className=""
            country={null}
            disabled={false}
            fixtures={Array []}
            getSuggestLabel={[Function]}
            googleMaps={null}
            highlightMatch={true}
            ignoreTab={false}
            initialValue=""
            inputClassName=""
            location={null}
            maxFixtures={10}
            minLength={1}
            onActivateSuggest={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onSuggestNoResults={[Function]}
            onSuggestSelect={[Function]}
            onUpdateSuggests={[Function]}
            placeholder="Write a location"
            queryDelay={250}
            radius={null}
            renderSuggestItem={null}
            required={true}
            skipSuggest={[Function]}
            style={
              Object {
                "input": Object {},
                "suggestItem": Object {},
                "suggests": Object {},
              }
            }
            title="Lorem ipsum"
            types={null}
          />
        </div>,
      ],
      "className": "locationInputWrapper",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Lorem ipsum",
          "className": "title",
        },
        "ref": null,
        "rendered": "Lorem ipsum",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Geosuggest
            autoActivateFirstSuggest={true}
            bounds={null}
            className=""
            country={null}
            disabled={false}
            fixtures={Array []}
            getSuggestLabel={[Function]}
            googleMaps={null}
            highlightMatch={true}
            ignoreTab={false}
            initialValue=""
            inputClassName=""
            location={null}
            maxFixtures={10}
            minLength={1}
            onActivateSuggest={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onSuggestNoResults={[Function]}
            onSuggestSelect={[Function]}
            onUpdateSuggests={[Function]}
            placeholder="Write a location"
            queryDelay={250}
            radius={null}
            renderSuggestItem={null}
            required={true}
            skipSuggest={[Function]}
            style={
              Object {
                "input": Object {},
                "suggestItem": Object {},
                "suggests": Object {},
              }
            }
            title="Lorem ipsum"
            types={null}
          />,
          "className": "locationInput",
        },
        "ref": "geosuggestWrapper",
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoActivateFirstSuggest": true,
            "bounds": null,
            "className": "",
            "country": null,
            "disabled": false,
            "fixtures": Array [],
            "getSuggestLabel": [Function],
            "googleMaps": null,
            "highlightMatch": true,
            "ignoreTab": false,
            "initialValue": "",
            "inputClassName": "",
            "location": null,
            "maxFixtures": 10,
            "minLength": 1,
            "onActivateSuggest": [Function],
            "onBlur": [Function],
            "onChange": [Function],
            "onFocus": [Function],
            "onSuggestNoResults": [Function],
            "onSuggestSelect": [Function],
            "onUpdateSuggests": [Function],
            "placeholder": "Write a location",
            "queryDelay": 250,
            "radius": null,
            "renderSuggestItem": null,
            "required": true,
            "skipSuggest": [Function],
            "style": Object {
              "input": Object {},
              "suggestItem": Object {},
              "suggests": Object {},
            },
            "title": "Lorem ipsum",
            "types": null,
          },
          "ref": "geosuggest",
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="title"
          >
            Lorem ipsum
          </div>,
          <div
            className="locationInput"
          >
            <Geosuggest
              autoActivateFirstSuggest={true}
              bounds={null}
              className=""
              country={null}
              disabled={false}
              fixtures={Array []}
              getSuggestLabel={[Function]}
              googleMaps={null}
              highlightMatch={true}
              ignoreTab={false}
              initialValue=""
              inputClassName=""
              location={null}
              maxFixtures={10}
              minLength={1}
              onActivateSuggest={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onSuggestNoResults={[Function]}
              onSuggestSelect={[Function]}
              onUpdateSuggests={[Function]}
              placeholder="Write a location"
              queryDelay={250}
              radius={null}
              renderSuggestItem={null}
              required={true}
              skipSuggest={[Function]}
              style={
                Object {
                  "input": Object {},
                  "suggestItem": Object {},
                  "suggests": Object {},
                }
              }
              title="Lorem ipsum"
              types={null}
            />
          </div>,
        ],
        "className": "locationInputWrapper",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Lorem ipsum",
            "className": "title",
          },
          "ref": null,
          "rendered": "Lorem ipsum",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Geosuggest
              autoActivateFirstSuggest={true}
              bounds={null}
              className=""
              country={null}
              disabled={false}
              fixtures={Array []}
              getSuggestLabel={[Function]}
              googleMaps={null}
              highlightMatch={true}
              ignoreTab={false}
              initialValue=""
              inputClassName=""
              location={null}
              maxFixtures={10}
              minLength={1}
              onActivateSuggest={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onSuggestNoResults={[Function]}
              onSuggestSelect={[Function]}
              onUpdateSuggests={[Function]}
              placeholder="Write a location"
              queryDelay={250}
              radius={null}
              renderSuggestItem={null}
              required={true}
              skipSuggest={[Function]}
              style={
                Object {
                  "input": Object {},
                  "suggestItem": Object {},
                  "suggests": Object {},
                }
              }
              title="Lorem ipsum"
              types={null}
            />,
            "className": "locationInput",
          },
          "ref": "geosuggestWrapper",
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "autoActivateFirstSuggest": true,
              "bounds": null,
              "className": "",
              "country": null,
              "disabled": false,
              "fixtures": Array [],
              "getSuggestLabel": [Function],
              "googleMaps": null,
              "highlightMatch": true,
              "ignoreTab": false,
              "initialValue": "",
              "inputClassName": "",
              "location": null,
              "maxFixtures": 10,
              "minLength": 1,
              "onActivateSuggest": [Function],
              "onBlur": [Function],
              "onChange": [Function],
              "onFocus": [Function],
              "onSuggestNoResults": [Function],
              "onSuggestSelect": [Function],
              "onUpdateSuggests": [Function],
              "placeholder": "Write a location",
              "queryDelay": 250,
              "radius": null,
              "renderSuggestItem": null,
              "required": true,
              "skipSuggest": [Function],
              "style": Object {
                "input": Object {},
                "suggestItem": Object {},
                "suggests": Object {},
              },
              "title": "Lorem ipsum",
              "types": null,
            },
            "ref": "geosuggest",
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
